modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs

```diff
@@ -215,5 +216,36 @@ ISymbol exportedMemberSymbol
                 location,
                 location?.SourceTree?.FilePath));
         }
+
+        public static readonly DiagnosticDescriptor GenericTypeArgumentMustBeVariantRule =
+            new DiagnosticDescriptor(id: "GD9001", // TODO: Not sure what ID to use but it needs to be this format, otherwise it throws an exception
+                title: "The generic type argument must be a Variant type",
+                messageFormat: "The generic type argument must be a Variant type: {0}",
+                category: "Usage",
+                DiagnosticSeverity.Error,
+                isEnabledByDefault: true,
+                "The generic type argument must be a Variant type. Use a Variant type as the generic type argument.");
+
+        public static void ReportGenericTypeArgumentMustBeVariant(
+            SyntaxNodeAnalysisContext context,
+            SyntaxNode typeArgumentSyntax,
+            ISymbol typeArgumentSymbol)
+        {
+            string message = "The generic type argument " +
+                            $"must be a Variant type: '{typeArgumentSymbol.ToDisplayString()}'";
+
+            string description = $"{message}. Use a Variant type as the generic type argument.";
+
+            context.ReportDiagnostic(Diagnostic.Create(
+                new DiagnosticDescriptor(id: "GD9001", // TODO: Not sure what ID to use but it needs to be this format, otherwise it throws an exception
```

Comment:
I tried `GODOT-GDXXXX` but it was throwing an exception:

```
CSC : warning AD0001: Analyzer 'Godot.SourceGenerators.MustBeVariantAnalyzer' threw an exception of type 'System.ArgumentException' with message 'Reported diagnostic has an ID 'GODOT-GD9001', which is not a valid identifier. (Parameter 'diagnostic')'.
```

Also, did you mean `GODOT-GDXXXX` or `GODOT-GXXXX`? Because currently source generators use a diagnostic Id without the **D**.

Reply:
Damn, then the source generator's diagnostics must have invalid IDs as well. I can't find any documentation about the format. We should change them before publishing.

> Also, did you mean `GODOT-GDXXXX` or `GODOT-GXXXX`? Because currently source generators use a diagnostic Id without the **D**.

`GODOT-GXXXX`, sorry. But it doesn't matter anymore :smile:

Reply:
I also couldn't find any documentation about the format, but looking at the source code it seems the id needs to be a valid unicode identifier (see https://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs,66) so I guess it must be the dash (`-`).

Interestingly enough, this validation only seems to be done in the analyzer's `AnalysisContext.ReportDiagnostic` method and not in the `GeneratorExecutionContext.ReportDiagnostic` method so diagnostics reported by source generators don't throw and show up in the IDE but diagnostics reported by an analyzer throw and don't show up in the IDE.

Reply:
It seems `_` is supported, so maybe `GODOT_GXXXX`, or better `GODOT_XXXX` while we are at it. I don't know why I ever went with that G prefix.

Reply:
Microsoft seems to use 2-3 letters and 4 digits (e.g.: `CA1051`, `IDE0067`) with no separator so I slightly prefer that but I don't have a strong opinion. I have pushed an update with `GODOT_XXXX` but since we'll probably want to change the source generator's diagnostics to be consistent we can decide then.

Reply:
I'm fine with either `GODOT_XXXX` or `GDXXXX`.

Reply:
I'll rename them to `GDXXXX` then.